File: models/conversation.py
"""会话历史的 ORM:会话 + 消息两张表(建在 db_wenshu,与 users/upload_datasets 同库)。

表结构:
  conversations
    id           主键
    user_id      归属用户(所有查询强制按它过滤 → 用户间历史天然隔离)
    source       'db'(主图问数) / 'dataset'(Excel 数据集问数)
    dataset_id   仅 source='dataset' 时有值,指向 upload_datasets.id
    title        会话标题(默认取首个问题截断,可改名)
    created_at / updated_at  updated_at 每次新消息时刷新,列表按它倒序

  messages
    id              主键
    conversation_id 所属会话(index)
    role            'user' / 'assistant'
    content         用户问题文本(role='user' 用)
    payload         JSON,助手回复的完整渲染结构
                    (sql / result / chart_config / interpretation / guide_queries / status / steps),
                    打开历史即原样重现,无需重跑 SQL
    created_at
"""
from datetime import datetime

from sqlalchemy import JSON, DateTime, Index, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from models import Base


class ConversationMySQL(Base):
    __tablename__ = "conversations"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True, comment="会话编号")
    user_id: Mapped[str] = mapped_column(String(64), index=True, comment="归属用户")
    source: Mapped[str] = mapped_column(String(16), default="db", comment="db / dataset")
    dataset_id: Mapped[int | None] = mapped_column(Integer, nullable=True, comment="数据集会话才有,指向 upload_datasets")
    # db 会话绑定的数据源(指向 meta 库 datasource.id);dataset 会话为空。
    # 删除数据源时按它连带清理会话;问数列表按它隔离(切到哪个源只看哪个源的历史)。
    datasource_id: Mapped[str | None] = mapped_column(String(64), nullable=True, comment="db 会话才有,指向 datasource.id")
    title: Mapped[str] = mapped_column(String(255), default="新对话", comment="会话标题")
    # 用 Python 端 default/onupdate(本地时间)而非仅 server_default(DB 端,容器多为 UTC),
    # 避免新建/更新时间与全项目 datetime.now() 不一致,导致前端按本地日期分组错位。
    # server_default 保留,仅作建表 DDL / 裸 SQL 插入的兜底。
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.now, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.now,
        server_default=func.now(),
        onupdate=datetime.now,
    )

    # 列表查询:数据集按 (user_id, source, dataset_id);问数按 (user_id, source, datasource_id)
    __table_args__ = (
        Index("idx_user_source_dataset", "user_id", "source", "dataset_id"),
        Index("idx_user_source_ds", "user_id", "source", "datasource_id"),
    )


class MessageMySQL(Base):
    __tablename__ = "messages"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True, comment="消息编号")
    conversation_id: Mapped[int] = mapped_column(Integer, index=True, comment="所属会话")
    role: Mapped[str] = mapped_column(String(16), comment="user / assistant")
    # 用户问题文本(assistant 消息此字段为空,内容在 payload)
    content: Mapped[str | None] = mapped_column(Text, nullable=True, comment="用户问题文本")
    # 助手回复的完整渲染结构(MySQL 8+ 原生 JSON)
    payload: Mapped[dict | None] = mapped_column(JSON, nullable=True, comment="助手回复渲染数据")
    # 同上:本地时间,与 conversations 保持一致
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.now, server_default=func.now())

File: wenshu-frontend/src/services/conversation.ts
import axios from 'axios'

import { getToken, redirectToLogin } from '@/lib/authToken'
import type { AgentReplyMessage, ChartConfig, ChatMessage } from '@/types/agent'

// 会话历史 REST(非流式),复用与 dataset.ts 一致的 Bearer 鉴权拦截器。
const convApi = axios.create({
  baseURL: import.meta.env.VITE_API_BASE_URL ?? '',
})

convApi.interceptors.request.use((config) => {
  const token = getToken()
  if (token) {
    config.headers.Authorization = `Bearer ${token}`
  }
  return config
})

convApi.interceptors.response.use(
  (res) => res,
  (error) => {
    if (axios.isAxiosError(error) && error.response?.status === 401) {
      redirectToLogin()
    }
    return Promise.reject(error)
  },
)

export type ConversationSource = 'db' | 'dataset'

export interface ConversationBrief {
  id: number
  source: ConversationSource
  dataset_id: number | null
  datasource_id: string | null
  title: string
  created_at: string | null
  updated_at: string | null
}

interface StoredMessage {
  id: number
  role: 'user' | 'assistant'
  content: string | null
  payload: Record<string, unknown> | null
  created_at: string | null
}

interface ConversationDetail extends ConversationBrief {
  messages: StoredMessage[]
}

// 新建一个空白会话(用户起名);返回新会话。问数会话绑数据源(datasourceId)。
export async function createConversation(
  source: ConversationSource,
  title: string,
  datasetId?: number,
  datasourceId?: string,
): Promise<ConversationBrief> {
  const { data } = await convApi.post<ConversationBrief>('/conversations', {
    source,
    title,
    dataset_id: source === 'dataset' ? (datasetId ?? null) : null,
    datasource_id: source === 'db' ? (datasourceId ?? null) : null,
  })
  return data
}

// 列表:问数传 source='db' + datasourceId(按数据源隔离);数据集传 source='dataset' + datasetId
export async function listConversations(
  source: ConversationSource,
  datasetId?: number,
  datasourceId?: string,
): Promise<ConversationBrief[]> {
  const params: Record<string, string | number> = { source }
  if (source === 'dataset' && datasetId != null) params.dataset_id = datasetId
  if (source === 'db' && datasourceId) params.datasource_id = datasourceId
  const { data } = await convApi.get<ConversationBrief[]>('/conversations', { params })
  return data
}

// 详情:把后端存的消息转成前端可直接渲染的 ChatMessage[]
export async function getConversationMessages(conversationId: number): Promise<ChatMessage[]> {
  const { data } = await convApi.get<ConversationDetail>(`/conversations/${conversationId}`)
  return data.messages.map(toChatMessage)
}

export async function renameConversation(conversationId: number, title: string): Promise<void> {
  await convApi.patch(`/conversations/${conversationId}`, { title })
}

export async function deleteConversation(conversationId: number): Promise<void> {
  await convApi.delete(`/conversations/${conversationId}`)
}

// 把按需生成的 chart_config 回写到某条 assistant 消息,落进历史(重开会话原样重现图表/提示)。
export async function persistMessageChart(
  conversationId: number,
  messageId: number,
  chartConfig: ChartConfig | null,
): Promise<void> {
  await convApi.patch(`/conversations/${conversationId}/messages/${messageId}/chart`, {
    chart_config: chartConfig,
  })
}

// 存储消息 → 前端 ChatMessage。assistant 的 payload 形状即 AgentReplyMessage(缺 id/role),补齐即可原样渲染。
function toChatMessage(m: StoredMessage): ChatMessage {
  if (m.role === 'user') {
    return { id: `hist-${m.id}`, role: 'user', content: m.content ?? '' }
  }
  const p = (m.payload ?? {}) as Partial<AgentReplyMessage>
  return {
    id: `hist-${m.id}`,
    // 后端消息 id:历史里再次「生成图表」时据它回写落库
    dbId: m.id,
    role: 'assistant',
    steps: p.steps ?? [],
    result: p.result ?? [],
    chartConfig: p.chartConfig ?? null,
    interpretation: p.interpretation ?? null,
    sql: p.sql ?? null,
    guideQueries: p.guideQueries ?? [],
    // 意图节点改写后的自包含问题:历史里再次出图/报告/归因时作数据问题(老数据无此字段则回退)
    standaloneQuestion: p.standaloneQuestion ?? null,
    // 扇出风险标记 + 文案:历史回放时重现警告样式(老数据无此字段则按非扇出处理)
    fanout: p.fanout ?? false,
    fanoutMessage: p.fanoutMessage ?? null,
    status: p.status ?? 'success',
  }
}
